timesketch/api/v1/resources/archive.py

```diff
@@ -462,58 +478,144 @@ def _unarchive_sketch(self, sketch: Sketch):
+            if not search_index.timelines:
+                logger.warning(
+                    "SearchIndex %s (ID: %s) has no associated timelines in DB. "
+                    "Skipping OpenSearch closure check.",
+                    search_index.index_name,
+                    search_index.id,
+                )
+                all_associated_timelines_archived = (
+                    False  # Cannot confirm, so don't archive OS index
+                )
 
-        # TODO (kiddi): Move this to lib/datastores/opensearch.py.
-        if indexes_to_close:
-            try:
-                self.datastore.client.indices.close(",".join(indexes_to_close))
-            except opensearchpy.NotFoundError:
-                logger.error(
-                    "Unable to close indices, not found: {:s}".format(
-                        ",".join(indexes_to_close)
+            for associated_timeline in search_index.timelines:
+                if associated_timeline.get_status.status != "archived":
+                    all_associated_timelines_archived = False
+                    logger.info(
+                        "SearchIndex %s (ID: %s) will not be closed/archived in DB "
+                        "because associated timeline %s status is '%s'.",
+                        search_index.index_name,
+                        search_index.id,
+                        associated_timeline.id,
+                        associated_timeline.get_status.status,
                     )
-                )
+                    break
+
+            if all_associated_timelines_archived:
+                # If all associated timelines are archived,
+                # attempt to close the OS index and then update
+                # the SearchIndex DB status.
+                original_si_status = search_index.get_status.status
+                try:
+                    logger.info(
+                        "Attempting to close OpenSearch index:"
+                        " %s (DB ID: %s) for sketch %s.",
+                        search_index.index_name,
+                        search_index.id,
+                        sketch.id,
+                    )
+                    self.datastore.client.indices.close(index=search_index.index_name)
+                    search_index.set_status(
+                        status="archived"
+                    )  # Set DB status only after successful close
+                    logger.info(
+                        "Successfully closed OpenSearch index: %s."
+                        " SearchIndex DB object "
+                        "(ID: %s) status updated to 'archived'.",
+                        search_index.index_name,
+                        search_index.id,
+                    )
+                except opensearchpy.exceptions.NotFoundError:
+                    # If index not found in OpenSearch, it's effectively gone/closed.
+                    search_index.set_status(status="archived")
```

Comment:
Is archived the correct status here? It implies that it can be unarchived, but if the index does not exist on the cluster this would not work. Maybe "deleted" or "fail" is better?

Reply:
oh that is a very good point, changed